packet/eip_controlplane_reconciliation.go

```diff
@@ -0,0 +1,131 @@
+package packet
+
+import (
+	"context"
+	"crypto/tls"
+	"net/http"
+	"time"
+
+	"errors"
+
+	"github.com/packethost/packngo"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog"
+)
+
+/*
+ controlPlaneEndpointManager checks the availability of an elastic IP for
+ the control plane and if it exists the reconciliation guarantees that it is
+ attached to a healthy control plane.
+
+ The general steps are:
+ 1. Check if the passed ElasticIP tags returns a valid Elastic IP via Packet API.
+ 2. If there is NOT an ElasticIP with those tags just end the reconciliation
+ 3. If there is an ElasticIP use the kubernetes client-go to check if it
+ returns a valid response
+ 4. If the response returned via client-go is good we do not need to do anything
+ 5. If the response if wrong or it terminated it means that the device behind
+ the ElasticIP is not working correctly and we have to find a new one.
+ 6. Ping the other control plane available in the cluster, if one of them work
+ assign the ElasticIP to that device.
+ 7. If NO Control Planes succeed, the cluster is unhealthy and the
+ reconciliation terminates without changing the current state of the system.
+*/
+type controlPlaneEndpointManager struct {
+	eipTag      string
+	instances   cloudInstances
+	deviceIPSrv packngo.DeviceIPService
+	ipResSvr    packngo.ProjectIPService
+	projectID   string
+	httpClient  *http.Client
+}
+
+func (m *controlPlaneEndpointManager) Reconciler() func(nodes []*v1.Node, remove bool) error {
+	return func(nodes []*v1.Node, remove bool) error {
+		klog.V(2).Info("controlPlaneEndpoint.reconcile: new reconciliation")
+		ipList, _, err := m.ipResSvr.List(m.projectID)
+		if err != nil {
+			return err
+		}
+		controlPlaneEndpoint := ipReservationByTags([]string{m.eipTag}, ipList)
+		if controlPlaneEndpoint == nil {
+			// IP NOT FOUND nothing to do here.
+			klog.Error("Elastic IP not found. Please verify you have one with the expected tag: %s", m.eipTag)
+			return err
+		}
+		klog.Infof("healthcheck elastic ip %s", controlPlaneEndpoint.Address)
+		req, err := http.NewRequest("GET", "https://"+controlPlaneEndpoint.Address+":6443/healthz", nil)
+		if err != nil {
+			klog.Errorf("healthcheck failed for control plane endpoint %s. err \"%s\". Checking the next one...", controlPlaneEndpoint.Address, err)
```

Comment:
I don't understand this error. If you got an error here, it isn't the health check that failed, but the creation of a new request. And what does "checking the next one" mean? The next what?